cpp-client/deephaven/client/src/server/server.cc

```diff
@@ -453,6 +506,78 @@ bool Server::processNextCompletionQueueItem() {
   return true;
 }
 
+namespace {
+class KeepAliveCallback final : public SFCallback<ConfigurationConstantsResponse> {
+public:
+  explicit KeepAliveCallback(std::shared_ptr<Server> server) : server_(std::move(server)) {}
+
+  void onSuccess(ConfigurationConstantsResponse resp) final {
+    auto expInt = extractExpirationInterval(resp);
+    if (expInt.has_value()) {
+      server_->setExpirationInterval(*expInt);
+    }
+  }
+
+  void onFailure(std::exception_ptr ep) final {
+    // TODO
+    std::cerr << "Keepalive failed\n";
+  }
+
+  std::shared_ptr<Server> server_;
+};
+}  // namespace
+
+void Server::sendKeepaliveMessages(const std::shared_ptr<Server> &self) {
+  while (true) {
+    if (!self->keepaliveHelper()) {
+      break;
+    }
+  }
+}
+
+bool Server::keepaliveHelper() {
+  // Wait for timeout or cancellation
+  {
+    std::unique_lock guard(mutex_);
+    std::chrono::system_clock::time_point now;
+    while (true) {
+      (void) condVar_.wait_until(guard, expirationTime_);
+      if (cancelled_) {
+        return false;
+      }
+      now = std::chrono::system_clock::now();
+      // Spurious condition variable wakeups can happen, and also we will get a deliberate wakeup
+      // in the hypothetical case where the server reduces the expiration interval and effectively
+      // moves the expiration time forward. So we only break out of the loop if we've actually
+      // passed the deadline.
+      if (now >= expirationTime_) {
+        break;
+      }
+    }
+
+    // Calculate new expiration time
+    expirationTime_ = now + expirationInterval_;
+  }
+
+  // Send a 'GetConfigurationConstants' as a handshake.
+  auto callback = std::make_shared<KeepAliveCallback>(shared_from_this());
+  getConfigurationConstantsAsync(std::move(callback));
```

Comment:
It seems we are sending getConfigurationConstant calls on a timer, in a fire-and-forget approach.  If the RPC fails (say network glitch or server unavailable or whatever), I believe we should detect that and retry.  Exactly how would depend on how we have deadlines and retries configured in the channel; I have some experience with how java gRPC can be configured for that, and is tricky (channel reconnection versus RPC retries and deadlines, and how they interact).
I am guessing we don't have deadlines configured for RPCs either, which will make gRPC retry forever (I think, at least in Java, and I say at least in Java because I've heard this stuff actually behaves differently in C++, maybe C++ is saner).
Let's have a quick chat.

Reply:
I looked a bit into Devin's latest changes to the java client and thought a bit more about this.  I thought of a simple solution here, that may be good enough and avoid the complexity of having to think about fine tuning of deadlines and retry attempts.  Leaving it here so that we can discuss it when we talk.

If I understand the code correctly, we are sending a keepalive in the middle of the expiration period; meaning if the period is 10 minutes, we send at 5.   It may be good enough to do something like "send every `max(30 seconds, period/10)`"